# Задачи
1. Импортируйте библиотеку pandas как pd. Загрузите датасет bookings.csv с разделителем ;. Проверьте размер таблицы, типы переменных, а затем выведите первые 7 строк, чтобы посмотреть на данные. 
2. Приведите названия колонок к нижнему регистру и замените пробелы на знак нижнего подчеркивания.
3. Пользователи из каких стран совершили наибольшее число успешных бронирований? Укажите топ-5.
4. На сколько ночей в среднем бронируют отели разных типов?
5. Иногда тип номера, полученного клиентом (assigned_room_type), отличается от изначально забронированного (reserved_room_type). Такое может произойти, например, по причине овербукинга. Сколько подобных наблюдений встретилось в датасете?
6. Проанализируйте даты запланированного прибытия. – На какой месяц чаще всего успешно оформляли бронь в 2016? Изменился ли самый популярный месяц в 2017?– Сгруппируйте данные по годам и проверьте, на какой месяц бронирования отеля типа City Hotel отменялись чаще всего в каждый из периодов
7. Посмотрите на числовые характеристики трёх переменных: adults, children и babies. Какая из них имеет наибольшее среднее значение?
8. Создайте колонку total_kids, объединив children и babies. Для отелей какого типа среднее значение переменной оказалось наибольшим?
9. Создайте переменную has_kids, которая принимает значение True, если клиент при бронировании указал хотя бы одного ребенка (total_kids), в противном случае – False. Посчитайте отношение количества ушедших пользователей к общему количеству клиентов, выраженное в процентах (churn rate). Укажите, среди какой группы показатель выше.

# Описание данных
Имеются следующие переменные:

* Hotel – тип отеля (City Hotel или Resort Hotel)  

* Is canceled – бронирование было отменено (1) или нет (0); не отмененное считается успешным

* Lead time – количество дней, прошедших между датой бронирования и датой прибытия  

* Arrival full date – полная дата прибытия

* Arrival date year – год прибытия  

* Arrival date month – месяц прибытия  

* Arrival date week number – номер недели прибытия

* Arrival date day of month – день прибытия

* Stays in weekend nights – количество выходных (суббота или воскресенье), которые гость забронировал для проживания в отеле

* Stays in week nights – количество дней (с понедельника по пятницу), которые гость забронировал для проживания в отеле

* Stays total nights – общее число забронированных ночей (сумма двух предыдущих колонок)

* Adults – число взрослых

* Children – число детей

* Babies – число младенцев 

* Meal – выбранный тип питания

* Country – страна происхождения клиента

* Reserved room type – тип зарезервированного номера

* Assigned room type – тип полученного номера (может отличаться от забронированного)

* Customer type – тип бронирования

* Reservation status – значение последнего статуса брони: Canceled - было отменено клиентом; Check-Out - клиент зарегистрировался, но уже покинул отель; No-Show - клиент не зарегистрировался и сообщил администрации отеля причину

* Reservation status date – дата обновления статуса


In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings("ignore")

%matplotlib inline

**1. Импортируем библиотеку pandas как pd. Прочитайте датасет bookings.csv с разделителем ;. Ссылка на датасет находится в переменной path_to_file.**

In [7]:
bookings = pd.read_csv('/Users/pavel/Desktop/karpov/2 МОДУЛЬ PYTHON +/2_bookings.csv', sep = ';')
bookings_head = bookings.head(7)
bookings_head

Hotel  Is Canceled  Lead Time arrival full date  Arrival Date Year  \
0  Resort Hotel            0        342        2015-07-01               2015   
1  Resort Hotel            0        737        2015-07-01               2015   
2  Resort Hotel            0          7        2015-07-01               2015   
3  Resort Hotel            0         13        2015-07-01               2015   
4  Resort Hotel            0         14        2015-07-01               2015   
5  Resort Hotel            0         14        2015-07-01               2015   
6  Resort Hotel            0          0        2015-07-01               2015   

  Arrival Date Month  Arrival Date Week Number  Arrival Date Day of Month  \
0               July                        27                          1   
1               July                        27                          1   
2               July                        27                          1   
3               July                        27                          1   
4               July                        27                          1   
5               July                        27                          1   
6               July                        27                          1   

   Stays in Weekend nights  Stays in week nights  ...  Adults  Children  \
0                        0                     0  ...       2       0.0   
1                        0                     0  ...       2       0.0   
2                        0                     1  ...       1       0.0   
3                        0                     1  ...       1       0.0   
4                        0                     2  ...       2       0.0   
5                        0                     2  ...       2       0.0   
6                        0                     2  ...       2       0.0   

   Babies  Meal Country Reserved Room Type Assigned room type customer type  \
0       0    BB     PRT                  C                  C     Transient   
1       0    BB     PRT                  C                  C     Transient   
2       0    BB     GBR                  A                  C     Transient   
3       0    BB     GBR                  A                  A     Transient   
4       0    BB     GBR                  A                  A     Transient   
5       0    BB     GBR                  A                  A     Transient   
6       0    BB     PRT                  C                  C     Transient   

  Reservation Status Reservation status_date  
0          Check-Out              2015-07-01  
1          Check-Out              2015-07-01  
2          Check-Out              2015-07-02  
3          Check-Out              2015-07-02  
4          Check-Out              2015-07-03  
5          Check-Out              2015-07-03  
6          Check-Out              2015-07-03  

[7 rows x 21 columns]

**2. Посмотрите на размеры таблицы. Сколько всего в ней столбцов?**

In [8]:
bookings_head.shape[1]

21

Ответ: 21

**3. К какому типу относится большинство переменных?**

In [10]:
bookings_head.dtypes.value_counts()

object     10
int64      10
float64     1
Name: count, dtype: int64


* **int64**
* **object**

**Приведем названия колонок к нормальному виду.**

Например: Is Canceled -> is_canceled

In [14]:
bookings = bookings.rename(columns = lambda col : col.replace(" ", '_').lower())
bookings.head()

hotel  is_canceled  lead_time arrival_full_date  arrival_date_year  \
0  Resort Hotel            0        342        2015-07-01               2015   
1  Resort Hotel            0        737        2015-07-01               2015   
2  Resort Hotel            0          7        2015-07-01               2015   
3  Resort Hotel            0         13        2015-07-01               2015   
4  Resort Hotel            0         14        2015-07-01               2015   

  arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
0               July                        27                          1   
1               July                        27                          1   
2               July                        27                          1   
3               July                        27                          1   
4               July                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  ...  adults  children  \
0                        0                     0  ...       2       0.0   
1                        0                     0  ...       2       0.0   
2                        0                     1  ...       1       0.0   
3                        0                     1  ...       1       0.0   
4                        0                     2  ...       2       0.0   

   babies  meal country reserved_room_type assigned_room_type customer_type  \
0       0    BB     PRT                  C                  C     Transient   
1       0    BB     PRT                  C                  C     Transient   
2       0    BB     GBR                  A                  C     Transient   
3       0    BB     GBR                  A                  A     Transient   
4       0    BB     GBR                  A                  A     Transient   

  reservation_status reservation_status_date  
0          Check-Out              2015-07-01  
1          Check-Out              2015-07-01  
2          Check-Out              2015-07-02  
3          Check-Out              2015-07-02  
4          Check-Out              2015-07-03  

[5 rows x 21 columns]

**Пользователи из каких стран совершили наибольшее число успешных бронирований? Бронирование считается успешным, если в дальнейшем не было отменено (переменная is_canceled).

In [76]:
bookings.query("is_canceled == 0").groupby('country', as_index = False).agg({"is_canceled" : "size"}).\
    sort_values('is_canceled', ascending = False).head()

country  is_canceled
125     PRT        21071
57      GBR         9676
54      FRA         8481
50      ESP         6391
42      DEU         6069

* **ESP**
* PLW
* IRL
* RUS
* **GBR**
* GUY
* MRT
* USA
* **PRT**
* **DEU**
* **FRA**

**На сколько ночей (stays_total_nights)  в среднем бронируют отели типа City Hotel? Resort Hotel?**

In [44]:
bookings.groupby('hotel', as_index = False).agg({"stays_total_nights": "mean"}).round(2)

hotel  stays_total_nights
0    City Hotel                2.98
1  Resort Hotel                4.32

**Иногда тип номера, присвоенного клиенту (assigned_room_type), отличается от изначально забронированного (reserved_room_type). Такое может произойти, например, по причине овербукинга.**

**Сколько подобных наблюдений встретилось в датасете?**

*отмена бронирования также считается

In [49]:
bookings.query("assigned_room_type != reserved_room_type").hotel.count()

14917

Ответ: 14917

**Теперь проанализируем даты запланированного прибытия (arrival_date_year).**

**На какой месяц чаще всего оформляли бронь в 2016 году? Изменился ли самый популярный месяц в 2017?**

In [66]:
bookings.groupby(['arrival_date_year']).arrival_date_month.agg(pd.Series.mode)

arrival_date_year
2015    September
2016      October
2017          May
Name: arrival_date_month, dtype: object

* **В 2016 чаще всего бронировали отель на октябрь**
* **В 2017 чаще всего оформляли бронирование на май**

**Сгруппируем данные по годам, а затем проверьте, на какой месяц (arrival_date_month) бронирования отеля типа City Hotel отменялись чаще всего в 2015? 2016? 2017?**

In [74]:
bookings.query("hotel == 'City Hotel' and is_canceled == 1").groupby('arrival_date_year').\
    arrival_date_month.agg(pd.Series.mode)

arrival_date_year
2015    September
2016      October
2017          May
Name: arrival_date_month, dtype: object

2015 => September

2016 => October

2017 => May

**Посмотрим на числовые характеристики трёх колонок: adults, children и babies. Какая из них имеет наибольшее среднее значение?**

In [80]:
bookings[['adults', 'children', 'babies']].mean()

adults      1.856403
children    0.103890
babies      0.007949
dtype: float64

* babies
* **adults**
* children

**Создадим колонку total_kids, объединив столбцы children и babies. Для отелей какого типа среднее значение переменной оказалось наибольшим?**

* City hotel – отель находится в городе
* Resort hotel – отель курортный

In [90]:
bookings['total_kids'] = bookings.children + bookings.babies
bookings.groupby('hotel', as_index = False).agg({"total_kids":'mean'}).round(2)

hotel  total_kids
0    City Hotel        0.10
1  Resort Hotel        0.14

Ответ: `0.14`

**Не все бронирования завершились успешно (is_canceled), поэтому можно посчитать, сколько клиентов было потеряно в процессе. Иными словами, посчитать метрику под названием Churn Rate.**

**Churn rate (отток, коэффициент оттока) – это процент подписчиков (например, на push-уведомления от сайта), которые отписались от канала коммуникации, отказались от услуг сервиса в течение определенного периода времени. Иными словами, представляет собой отношение количества ушедших пользователей к общему количеству пользователей, выраженное в процентах.**

**В нашем случае Churn Rate - это процент клиентов, которые отменили бронирование. Давайте посмотрим, как эта метрика связана с наличием детей у клиентов!**

**Создадим переменную has_kids, которая принимает значение True, если клиент при бронировании указал хотя бы одного ребенка (total_kids), в противном случае – False. Далее проверим, среди какой группы пользователей показатель оттока выше.** 

In [91]:
bookings['has_kids'] = bookings.total_kids > 0

In [107]:
bookings[['has_kids', 'is_canceled']].value_counts() / bookings.has_kids.value_counts() * 100

has_kids  is_canceled
False     0              62.778717
          1              37.221283
True      0              65.077154
          1              34.922846
Name: count, dtype: float64

Ответ: `37.22%`